### PR TITLE
Anonymous tracking with Segment.com

### LINKED ISSUE #91: Tracking
I'm wondering whether it'd be worth adding some kind of internal tracking that gets sent to directly back to us, to see how often something happens.

We'd only need to track:

- Installs
- Components; added, edited, deleted
- Incidents; added, edited, deleted
- Notifications

Then we'd have enough data to use as "we've had x installs" etc.

Of course, there are things like:

- Security
- Privacy

That we'd need to take care of.

### PR DESCRIPTION
**Pull request by jbrooksuk**
Closes #91.

We track a few things, but could easily add more, plus it's all anonymous!

# Things to do:

- [x] Setting of the Segment.com `write_key` should be based on an API result we fetch and store somewhere, this allows us to revoke it at any time from all status pages.
- [x] What happens if the API key is completely wrong and the new one is not installed? Does Segment lose its shit? **Nothing. It just doesn't work - this is good.**

### CODE REVIEW

File: app/config/app.php [hunk 1]
```diff
@@ -137,6 +137,7 @@
         'McCool\LaravelAutoPresenter\LaravelAutoPresenterServiceProvider',
         'PragmaRX\Google2FA\Vendor\Laravel\ServiceProvider',
         'Fideloper\Proxy\ProxyServiceProvider',
+        'CachetHQ\Segment\SegmentServiceProvider',
```
Comment (GrahamCampbell) [inline app/config/app.php]: Needs to be alphabetically sorted,

File: app/start/global.php [hunk 2]
```diff
@@ -61,3 +63,9 @@
 App::down(function () {
     return Response::make("Be right back!", 503);
 });
+
+/**
+ * Segment.com tracking.
+ * Since every user is identified under one site, we can "identify" them here.
```
Comment (GrahamCampbell) [inline app/start/global.php]: We need another new line before the start of the long description.

File: app/start/global.php [hunk 3]
```diff
@@ -61,3 +63,9 @@
...
+ */
+segment_identify(Config::get('app.key'));
```
Comment (GrahamCampbell) [inline app/start/global.php]: This means we're setting this up on every request. Probably not what we want to do.

File: app/config/app.php (hunk 1 above)
Comment (jbrooksuk) [inline app/config/app.php]: Ok :)

File: app/start/global.php (hunk 2 above)
Comment (jbrooksuk) [inline app/start/global.php]: Gotcha.

File: app/start/global.php (hunk 3 above)
Comment (jbrooksuk) [inline app/start/global.php]: What I could do is move this line into the `segment_` helpers?

**Comment by jbrooksuk** [comment]
I've setup Keen.io integration which is much better for our usage.

**Comment by jbrooksuk** [comment]
@GrahamCampbell I'm storing the write key in a private Gist file. It's nothing bad ass, but since we have to release it at some point, it's the easiest way.

What do you think is the best way for us to load this? I'm thinking of having it store as a setting, then if the `created_at` is more than a day old, check again. We wouldn't lose too much data if it then had to be changed.

**Comment by jbrooksuk** [comment]
CC @joecohens (see above).

**Comment by jbrooksuk** [comment]
1. Needs rebasing.
2. If we can figure out how to remotely fetch the write key every few hours, then update the setting, we're able to merge.

**Comment by jbrooksuk** [comment]
Rebased :)

File: src/Providers/LoadConfigServiceProvider.php [hunk 4]
```diff
@@ -26,6 +26,11 @@ public function boot()
             // Don't throw any errors, we may not be setup yet.
         }
 
+        // Set the Segment.com settings.
+        $repository = $this->app->make('CachetHQ\Cachet\Segment\RepositoryInterface');
+        $segmentWriteKey = $repository->fetch();
+        $this->app->config->set('cachethq/segment::write_key', $segmentWriteKey);
```
Comment (joecohens) [inline src/Providers/LoadConfigServiceProvider.php]: Inline?

File: src/Segment/CacheRepository.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+use CachetHQ\Cachet\Facades\Setting;
+use CachetHQ\Cachet\Models\Setting as SettingModel;
+use Carbon\Carbon;
+
+class CacheRepository implements RepositoryInterface
+{
+    protected $repository;
+
+    public function __construct(RepositoryInterface $repository)
```
Comment (joecohens) [inline src/Segment/CacheRepository.php]: Comments

File: src/Providers/LoadConfigServiceProvider.php (hunk 4 above)
Comment (jbrooksuk) [inline src/Providers/LoadConfigServiceProvider.php]: ?

File: src/Segment/RepositoryInterface.php [hunk 6]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace CachetHQ\Cachet\Segment;
+
+interface RepositoryInterface
+{
+    public function fetch();
```
Comment (joecohens) [inline src/Segment/RepositoryInterface.php]: Comments

File: src/Providers/LoadConfigServiceProvider.php (hunk 4 above)
Comment (jbrooksuk) [inline src/Providers/LoadConfigServiceProvider.php]: Gotcha. Will fix.

File: src/Segment/RepositoryInterface.php (hunk 6 above)
Comment (jbrooksuk) [inline src/Segment/RepositoryInterface.php]: I knew I'd forgotten something :P

File: src/Providers/LoadConfigServiceProvider.php (hunk 4 above)
Comment (joecohens) [inline src/Providers/LoadConfigServiceProvider.php]: `$segmentRepository = $this->app->make('CachetHQ\Cachet\Segment\RepositoryInterface');`
`$this->app->config->set('cachethq/segment::write_key', $segmentRepository->fetch());`

**Comment by jbrooksuk** [comment]
Just needs a rebase now.

**Comment by joecohens** [comment]
@jbrooksuk Needed rebasing again lol.

**Comment by jbrooksuk** [comment]
I take it you fixed style CI?

**Comment by joecohens** [comment]
Yep I've fixed StyleCI, rebased and merged :)

**Pull request merged**